core/src/main/scala/org/bykn/bosatsu/DefRecursionCheck.scala

```diff
@@ -0,0 +1,251 @@
+package org.bykn.bosatsu
+
+import cats.data.{NonEmptyList, Validated, ValidatedNel}
+
+import cats.implicits._
+
+/**
+ * Recursion in bosatsu is only allowed on a substructural match
+ * of one of the parameters to the def. This strict rule, along
+ * with strictly finite data, ensures that all recursion terminates
+ *
+ * The rules are as follows:
```

Comment:
I would especially appreciate reviews to this, and any tests anyone would like to see added.